Power spectrum responses for SSC

RyoTerasawa opened the pull request:
In pyccl/pkresponse.py, I implemented the functions to calculate power spectrum responses to the super-survey modes which are responsible for super-sample covariance (SSC), based on [arXiv:2310.13330](https://arxiv.org/abs/2310.13330).
We approximate the power spectrum growth response to the super-survey modes by its growth response to the Hubble parameter.

For the galaxy-matter and galaxy-auto power spectra, the halo statistics (halo-matter/halo-auto power spectrum, mass function) are calculated by DarkEmulator.

RyoTerasawa committed:
add darkemulator.py

RyoTerasawa committed:
add darkemulator mass function

RyoTerasawa committed:
Linting

RyoTerasawa committed:
delete darkemulator.py

RyoTerasawa committed:
add dependency on dark_emulator

RyoTerasawa committed:
Update environment.yml

RyoTerasawa committed:
Merge pull request #1141 from LSSTDESC/master

Update to master

damonge commented:
@RyoTerasawa can I check what the status of this PR is?

RyoTerasawa commented:
I am waiting for someone to review the code. I haven't asked for a specific person yet, but maybe I should ask a member of the MCPCov group to review it. I am happy to explain the details if needed. I am also working on writing the unit test and benchmark test to be included in this PR.

RyoTerasawa committed:
Merge remote-tracking branch 'origin/master' into SSC

RyoTerasawa committed:
Merge pull request #1166 from LSSTDESC/master

Update to v3.0.1

carlosggarcia commented:
@RyoTerasawa @YueNan-c, maybe you want to also implement the DarkEmulator here: https://github.com/LSSTDESC/CCL/blob/master/pyccl/emulators/cosmicemu_pk.py.

Are the unit tests and benchmark ready?

YueNan-c committed:
Local updates

YueNan-c committed:
Local updates

Unit test updates

YueNan-c committed:
Update test_pkresponse.py

YueNan-c committed:
Add files via upload

YueNan-c committed:
Delete benchmarks/test_ssc_terasawa_accuracy.py

damonge commented:
@RyoTerasawa @carlosggarcia can I check what the status of this is?

damonge committed:
Merge branch 'master' into SSC

damonge committed:
try bumping python

damonge committed:
try george in conda

damonge committed:
Merge branch 'master' into SSC

RyoTerasawa committed:
add unit test and benchmark test

RyoTerasawa committed:
lint

RyoTerasawa commented:
@carlosggarcia @damonge 
Sorry for the late reply. I just added the unit tests and benchmark, and this pull request is ready to be reviewed.

carlosggarcia commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,795 @@
+        dpklin = pk2dlin.__call__(k_use, aa, cosmo, derivative=True)
+
+        # use the perturbation theory below kmin
+        T_h[k_use <= kmin] = 1
```
This can be moved out of the loop

RyoTerasawa replied:
I moved the line (T_h = 1) out of the loop.

carlosggarcia commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,795 @@
+        dpk[k_use <= kmin] = dpklin[k_use <= kmin]
+        dpk[k_use > kmin] = dpknl[k_use > kmin]
+
+        dpk12[ia, :] = pk * (1.0 + (26.0 / 21.0) * T_h - dpk / 3.0)
```
It'd be great to have a comment with the Eq number that this corresponds to (for future reference)

carlosggarcia replied:
Same with the other equations

carlosggarcia replied:
Thanks! Two questions now:
 - In Eq. 23 the first term is a 2, where does the 1 come from?
 - is dpk = d ln(k^3 Pmm) / d lnk?

RyoTerasawa replied:
Actually it is dpk = d ln(Pmm) / d lnk. 
`2 - (1/3) d ln(k^3 Pmm) / d lnk = 1 - (1/3) d ln(Pmm) / d lnk`
Eq. 23 is l.h.s. and the code corresponds to the r.h.s.

carlosggarcia commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,795 @@
+
+    kmin = 1e-2
+    for ia, aa in enumerate(a_arr):
+        pk = pk2d.__call__(k_use, aa, cosmo)
```
You don't need to use `__call__`, you can just call it, `pk2d(k_use, aa, cosmo)`

RyoTerasawa replied:
Replaced the all `__call__` with `pk2d()` in `pkresponse.py`.

carlosggarcia commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,795 @@
+    hbf = halos.HaloBiasTinker10(mass_def=mass_def)
+
+    # dark emulator is valid for 0 =< z <= 1.48
+    if np.any(1.0 / a_arr - 1) > 1.5:
```
if the maximum is 1.48, check for 1.48, not 1.5

RyoTerasawa replied:
Modified the `pkresponse.py` to check for 1.48.

carlosggarcia commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,795 @@
+
+    # dark emulator is valid for 0 =< z <= 1.48
+    if np.any(1.0 / a_arr - 1) > 1.5:
+        print("dark emulator is valid for z={:.2f}<1.48")
```
hmm, what does `{:.2f}` do here? I think you have to remove that.

RyoTerasawa replied:
The `{:.2f}` was unnecessary. I removed that.

carlosggarcia commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,795 @@
+            / ng
+        )
+
+        bgE2 = (
```
@damonge , can this be done with the current methods of the halo model?

carlosggarcia replied:
Same with the other integrals

damonge replied:
I think so. @RyoTerasawa can you check:
https://github.com/LSSTDESC/CCL/blob/5273ca6eb76d7c92fd11cde42bf02f761e8c29cd/pyccl/halos/halo_model.py#L108
and 
https://github.com/LSSTDESC/CCL/blob/5273ca6eb76d7c92fd11cde42bf02f761e8c29cd/pyccl/halos/halo_model.py#L113
and
https://github.com/LSSTDESC/CCL/blob/5273ca6eb76d7c92fd11cde42bf02f761e8c29cd/pyccl/halos/halo_model.py#L118
?

RyoTerasawa replied:
I changed the `pkresponse.py`, and now all the integrals of the mass function are using `_integrate_over_mf` or `_integrate_over_mbf`.

carlosggarcia commented on benchmarks/test_pkresponse.py:
```diff
@@ -0,0 +1,143 @@
+# Construct the full path for each data file (z=0)
+k_data_path = os.path.join(data_directory_path, "k_h.npy")
+k_data_mm_path = os.path.join(data_directory_path, "k_h_mm.npy")
+Pmm_resp_data_path = os.path.join(data_directory_path, "Pmm_resp_z0.npy")
```
You have so much more data generated, at different redshifts. Why don't you use it?

RyoTerasawa replied:
I added the tests for higher redshifts, z ~ [0.5, 1.0, 1.5].

carlosggarcia reviewed (changes requested):
Ok. I first round of comments. Some suggestions:
- Add unit tests per function/method declared to be sure they are doing what they should
- I would unify the `.npy` files in a single `.npz` file. Also, it seems that you are only using the z=0 case, I'd be great if you can compare all redshift, since you have it.
- Add References to the equations in your paper for later reference (and to make my life easier when reviewing the PR)
- @damonge, there are integrals and biases computed that I think they might be obtained from the current CCL implementation. Can you have a look? The are in `pkresponse.py`, where I pinged you, too.

In general, it looks good, though! Good work!

damonge commented:
@carlosggarcia @RyoTerasawa can I check what the status of this PR is?

RyoTerasawa commented:
> @carlosggarcia @RyoTerasawa can I check what the status of this PR is?

I am currently working on addressing Carlos's comments.

RyoTerasawa committed:
Merge branch 'master' into SSC

RyoTerasawa committed:
Add References to the equations

RyoTerasawa commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,795 @@
+
+    omega_c = Omega_c * h**2
+    omega_b = Omega_b * h**2
+    omega_nu = 0.00064
```
This is because DarkEmulator is assuming m_nu = 0.06 eV.

RyoTerasawa committed:
switch to MassFuncNishimichi19

RyoTerasawa commented on pyccl/halos/hmfunc/darkemulator.py:
```diff
@@ -0,0 +1,65 @@
+    def _check_mass_def_strict(self, mass_def):
+        if isinstance(mass_def.Delta, str):
+            return True
+        elif int(mass_def.Delta) == 200:
```
https://github.com/LSSTDESC/CCL/pull/1138

RyoTerasawa commented on pyccl/halos/hmfunc/darkemulator.py:
```diff
@@ -0,0 +1,65 @@
+
+        omega_c = Omega_c * h**2
+        omega_b = Omega_b * h**2
+        omega_nu = 0.00064
```
Same as above, I switched to the DarkEmulator mass function class recently added to the main branch: #1138

RyoTerasawa commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,795 @@
+def Pmm_resp(
+    cosmo,
+    deltah=0.02,
+    extra_parameters={
```
I added dependency to `extra_parameters`.

RyoTerasawa committed:
add dependency to extra_parameters

RyoTerasawa committed:
make sure the cosmology object is exactly the same

RyoTerasawa commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,795 @@
+
+
+def set_hmodified_cosmology(cosmo, deltah):
+    Omega_c = cosmo["Omega_c"]
```
In the massive neutrinos' case, `m_nu` should be the same. I edited the function `set_hmodified_cosmology` to explicitly make sure the cosmology object is exactly same except for the `h`, by starting with the copy of the fiducial cosmology object through its keys.

carlosggarcia replied:
Great!

RyoTerasawa committed:
create pkresponse.npz

RyoTerasawa committed:
update benchmark

RyoTerasawa committed:
add documentation

RyoTerasawa commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,795 @@
+    return b2
+
+
+def darkemu_set_cosmology(emu, cosmo):
```
I made all the utility functions private. Also added documentation.

RyoTerasawa committed:
make utility functions private and add unit tests for them

RyoTerasawa commented on pyccl/tests/test_pkresponse.py:
I added a unit test for all functions defined in `pkresponse.py`.  For `set_hmodified_cosmology` the unit test checks that the output cosmologies are exactly the same as the input one, except for the h's.

RyoTerasawa committed:
update test_pkresponse.py

RyoTerasawa commented on pyccl/tests/test_pkresponse.py:
```diff
@@ -0,0 +1,46 @@
+
+    assert np.all(
+        np.isfinite(response)
+    ), "darkemu_Pgm_resp produced infinity values."
```
It was bit difficult for me to come up with good limits, so for now I just check the sign of the response (which depends on the HOD parameters).

RyoTerasawa committed:
lint

RyoTerasawa commented:
@carlosggarcia Thank you for the comments and suggestions.
I complete addressing these comments except for the integrals.

1. Added unit tests per function/method declared to be sure they are doing what they should
2. Unified the .npy files in a single .npz file and now the benchmark test compares all redshift.
3. Added References to the equations in the paper.
4. Have not addressed to the integrals in `pkresponse.py`

> @damonge, there are integrals and biases computed that I think they might be obtained from the current CCL implementation. Can you have a look? The are in pkresponse.py, where I pinged you, too.

damonge commented:
@RyoTerasawa I responded to Carlos' comment about the integrals. If possible, it'd be good if CCL does all of these consistently.

RyoTerasawa committed:
switched to halo model integral

RyoTerasawa commented:
@damonge Thank you for the suggestion. I switched to the integral method of `pyccl/halos/halo_model.py` and it helped to make the code bit simpler.

RyoTerasawa committed:
Merge pull request #1217 from LSSTDESC/master

Merge branch 'master' into SSC

RyoTerasawa commented:
@carlosggarcia, I complete addressing these comments. Although the build failed at the unit test.

RyoTerasawa committed:
rerun Github jobs

RyoTerasawa committed:
Merge remote-tracking branch 'origin/SSC' into SSC

RyoTerasawa committed:
change to Miniforge3

RyoTerasawa committed:
Update ci.yml

RyoTerasawa committed:
Merge pull request #1225 from LSSTDESC/master

Allowing for cross-correlations in HOD power spectrum. (#1224)

RyoTerasawa commented:
@carlosggarcia I believe this PR is ready for review. Whenever you have time, I would appreciate your feedback. Thank you!

carlosggarcia commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,803 @@
+        )
+
+        # use the perturbation theory below kmin
+        kmin = 1e-2  # [h/Mpc]
```
I see that you use this in different functions. It would be better if it could be set as a global quantity so that if you change it once, you change it for all

Also, the units of the k and Pk's in CCL are without little `h`. I think you will need to account for this

RyoTerasawa replied:
I made `khmin` a global quantity set at the first line of the `pkresponse.py`, while allowing users to set non default value via the argument of the functions. 

I corrected the units of k, and I believe I took into account the units correctly now.

carlosggarcia replied:
Good, thanks!

carlosggarcia commented on benchmarks/test_pkresponse.py:
```diff
@@ -0,0 +1,149 @@
+    assert np.allclose(
+        Pmm_resp_data[0, indx_mm],
+        generated_Pmm_resp[0],
+        atol=10 * Pmm_resp_err_data[0, indx_mm],
```
Aren't all the `atol` a bit too big? this is like 10 sigma, right? And you have others with even larger numbers.

RyoTerasawa replied:
I guess the error bars from the simulations are too strict, probably because we use the pair and fixed initial density method. Instead, I switched to `rtol`. 
The difference between the model and the simulations are typically few ten percents. Higher redshifts of Pgm, Pgg (especially Pgg) have worse accuracy due to the accuracy of `darkemulator` itself, but they are still a factor of difference at most.

carlosggarcia replied:
Okay, make sense. Can you add a comment in the tests explaining this? Otherwise we will forget in a few months.

RyoTerasawa replied:
I added above comments in the `benchmarks/test_pkrespose.py`.

carlosggarcia reviewed (changes requested):
A few more comment/questions. 

I didn't have time to go through all the equation in `pkresponse.py` but I guess that if the benchmark pass, they will be ok.

The most important comment is the one regarding the benchmarks. It seems to me that the threshold is to high (e.g. 60 sigma away of what you measure in the simulation)

RyoTerasawa committed:
Merge remote-tracking branch 'origin/SSC' into SSC

RyoTerasawa committed:
update benchmark

carlosggarcia commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,805 @@
+        nP = nth_mat * np.array(Pth)
+
+        # Eq. A7
+        Pgm = integrate.romb(dprof_dlogM * nP, dx=dlogM, axis=0) / ng
```
I think this could be done with current halo calculator, @damonge?

I've commented those that I think might use current CCL integrators

carlosggarcia commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,805 @@
+        Pgg_2h_int = np.array(Pgg_2h_int)
+
+        # Eq. A12
+        _Pgg_2h = integrate.romb(
```
Maybe this one too? @damonge

carlosggarcia commented on pyccl/pkresponse.py:
```diff
@@ -0,0 +1,805 @@
+        G_prof = (
+            +2.0
+            / 3.0
+            * integrate.romb(
```
Maybe this too? @damonge

carlosggarcia reviewed (commented):
All looks good to me. Can you merge the master branch into this branch to see if that solves the issue with the unit tests? 
Once the unit test pass, I will approve.

I've left some questions for @damonge regarding the integrals.

RyoTerasawa committed:
add comments in benchmark

RyoTerasawa committed:
Merge pull request #1239 from LSSTDESC/master

merge from master branch

RyoTerasawa committed:
update pkresponse.py

RyoTerasawa committed:
Merge remote-tracking branch 'origin/SSC' into SSC

RyoTerasawa commented:
@carlosggarcia I figured out I should do `from .errors import warnings` rather than just `import warnings`. 
After fixing it, all the test are passed.

RyoTerasawa committed:
dealt with coveralls

RyoTerasawa committed:
update test_pkresponse.py

RyoTerasawa committed:
update pkresponse.py

RyoTerasawa committed:
update CCLWarning import

RyoTerasawa committed:
update pkresponse.py

RyoTerasawa committed:
Make wanings into ValueError

RyoTerasawa committed:
Merge pull request #1245 from LSSTDESC/master

Merge changes in master

RyoTerasawa committed:
lint

RyoTerasawa committed:
Merge remote-tracking branch 'origin/SSC' into SSC

carlosggarcia reviewed (changes requested):
I would re-add the lk_arr and a_arr options. @damonge can you check the comments above for the integrals?

RyoTerasawa committed:
re-add lk_arr and a_arr options

carlosggarcia commented on pyccl/pkresponse.py:
```diff
@@ -42,20 +42,20 @@ def Pmm_resp(
-        raise ValueError("you must provide arrays")
+    # Set k and a sampling from CCL parameters
+    if a_arr is None:
+        a_arr = get_pk_spline_a()
```
Thanks! I would use the same code that was there before, just in case there is a subtle difference that we are missing
```
    if lk_arr is None:
        status = 0
        nk = lib.get_pk_spline_nk(cosmo.cosmo)
        lk_arr, status = lib.get_pk_spline_lk(cosmo.cosmo, nk, status)
        check(status, cosmo=cosmo)
    if a_arr is None:
        status = 0
        na = lib.get_pk_spline_na(cosmo.cosmo)
        a_arr, status = lib.get_pk_spline_a(cosmo.cosmo, na, status)
        check(status, cosmo=cosmo)
```

carlosggarcia reviewed (changes requested):
Almost there

RyoTerasawa committed:
update spline

carlosggarcia reviewed (approved):
LGTM! I'd wait for @damonge to check whether some of the integrals are duplicating some existing functionality before merging.

damonge reviewed (changes requested):
I trust the functional contents of this PR are fine since they've been reviewed by  @carlosggarcia . I don't mind that the halo model integrals are carried out outside of the current halo model (we could incorporate this later if this module becomes more embedded in the pipeline). However, I have a few requests about the structure of the new module itself:

- First, change the name of the module to `pk_response` rather than `pkresponse`, for clarity.
- Second, we need better names for all the new functions, so they are easily identifiable. My suggestion is as follows:
  - Use a structure of the form `resp_<pk_kind>_<method>`, where `<pk_kind>` is `Pmm`, `Pgm` etc., and `<method>` is the method used to calculate the response.
  - With this in mind, the new functions you have added should read `resp_Pmm_hresponse` (for example), `resp_Pgm_darkemu` etc.

I also have an overarching comment: presumably we would like to use the contents of this PR to improve the way we compute the SSC covariances. However, right now the methods you have implemented are completely orphan, and cannot be used for this. The way things currently stand, the SSC contribution computed [here](https://github.com/LSSTDESC/CCL/blob/master/pyccl/halos/pk_4pt.py#L322) automatically assumes the PT approach to the power spectrum response.

Before we move forward, can I ask what the plans are for embedding the new responses in the existing SSC infrastructure? I'm asking because the structure of the code added by this PR should be influenced by how we will use this code later on.

RyoTerasawa committed:
renamed module and functions

RyoTerasawa commented:
@carlosggarcia @damonge Thank you for the through review. 

- I changed the name of the module to pk_response.
- I renamed the new functions as resp_Pmm_hresponse, resp_Pgm_darkemu,  and resp_Pgg_darkemu.
- I added the  documentation for resp_Pgm_darkemu,  and resp_Pgg_darkemu.
- Regarding the last point, the outputs of the pk_response module can be passed to `pyccl.tk3d.Tk3D` to make a `tk3d` object, and then  using the `tk3d` object as an input of `pyccl.cosmology.angular_cl_cov_SSC`, we can compute the SSC. I will write this interface in TJPCov and possibly in CCL too.

RyoTerasawa committed:
fix typo

carlosggarcia reviewed (commented):
Two missing arguments in the documentation.

Also, @damonge, we will need your approval

Commit:
add docs for khmin

YueNan-c committed:
Merge branch 'SSC' of https://github.com/LSSTDESC/CCL into SSC
